src/api/router/Router.ts

```diff
@@ -0,0 +1,115 @@
+import { NavigationView } from "tabris";
+import { ListLike, Mutation } from "../List";
+import { RouterMatcher } from "./RouterMatcher";
+import { RouterHistory, HistoryItem } from "./RouterHistory";
+import { Route } from "./Route";
+import { Constructor } from "../../internals/utils";
+
+export type RouterConfig = {
+  name: string;
+  route: Constructor<Route>
+};
+
+export type RouterProperties = {
+  navigationView: NavigationView,
+  routers?: ListLike<RouterConfig>,
+  defaultRoute?: HistoryItem,
+  history?: ListLike<HistoryItem>
+};
+
+export class Router<ItemType extends  HistoryItem = HistoryItem> {
+
+  private _navigationView: NavigationView;
+  private _routes: ListLike<RouterConfig>;
+
+  private _routerHistoryObserver: RouterHistory;
+  private _routerMatcher: RouterMatcher;
+
+  constructor({navigationView, routers, defaultRoute, history} : RouterProperties) {
+    this._navigationView = navigationView;
+    this.routes = routers || [];
+    this._routerHistoryObserver = new RouterHistory(this._handleHistoryChange);
+    this.history = history || [];
+    this._routerMatcher = new RouterMatcher(this);
+    if (defaultRoute) {
+      this.goTo(defaultRoute as ItemType);
+    }
+    this._navigationView.onRemoveChild(this._syncHistoryWithNavigationView.bind(this));
+  }
+
+  goTo(item: ItemType) {
+    this._routerHistoryObserver.push(item);
+  }
+
+  back() {
+    if (this._routerHistoryObserver.source.length === 0) {
+      throw new Error("Could not call back on empty history stack");
+    }
+    this._routerHistoryObserver.pop();
+  }
+
+  set history(value: ListLike<HistoryItem>) {
+    if (this._routerHistoryObserver.source === value) {
+      return;
+    }
+    this._routerHistoryObserver.source = value;
```

Comment:
I think we should create safe copies for setting an getting history as array. That way the user can't modify the array and cause it to be out of sync.